File: k9mail/src/main/java/com/fsck/k9/activity/EditIdentity.java
package com.fsck.k9.activity;

import android.os.Bundle;
import android.view.View;
import android.widget.CheckBox;
import android.widget.CompoundButton;
import android.widget.EditText;
import android.widget.LinearLayout;
import com.fsck.k9.Account;
import com.fsck.k9.Identity;
import com.fsck.k9.Preferences;
import com.fsck.k9.R;
import java.util.List;

public class EditIdentity extends K9Activity {

    public static final String EXTRA_IDENTITY = "com.fsck.k9.EditIdentity_identity";
    public static final String EXTRA_IDENTITY_INDEX = "com.fsck.k9.EditIdentity_identity_index";
    public static final String EXTRA_ACCOUNT = "com.fsck.k9.EditIdentity_account";

    private Account account;
    private Identity identity;
    private int identityIndex;
    private EditText descriptionView;
    private CheckBox signatureUse;
    private EditText signatureView;
    private LinearLayout signatureLayout;
    private EditText emailView;
//  private EditText mAlwaysBccView;
    private EditText nameView;
    private EditText replyTo;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        identity = (Identity)getIntent().getSerializableExtra(EXTRA_IDENTITY);
        identityIndex = getIntent().getIntExtra(EXTRA_IDENTITY_INDEX, -1);
        String accountUuid = getIntent().getStringExtra(EXTRA_ACCOUNT);
        account = Preferences.getPreferences(this).getAccount(accountUuid);

        if (identityIndex == -1) {
            identity = new Identity();
        }

        setContentView(R.layout.edit_identity);

        /*
         * If we're being reloaded we override the original account with the one
         * we saved
         */
        if (savedInstanceState != null && savedInstanceState.containsKey(EXTRA_IDENTITY)) {
            identity = (Identity)savedInstanceState.getSerializable(EXTRA_IDENTITY);
        }

        descriptionView = (EditText)findViewById(R.id.description);
        descriptionView.setText(identity.getDescription());

        nameView = (EditText)findViewById(R.id.name);
        nameView.setText(identity.getName());

        emailView = (EditText)findViewById(R.id.email);
        emailView.setText(identity.getEmail());

        replyTo = (EditText) findViewById(R.id.reply_to);
        replyTo.setText(identity.getReplyTo());

//      mAccountAlwaysBcc = (EditText)findViewById(R.id.bcc);
//      mAccountAlwaysBcc.setText(identity.getAlwaysBcc());

        signatureLayout = (LinearLayout)findViewById(R.id.signature_layout);
        signatureUse = (CheckBox)findViewById(R.id.signature_use);
        signatureView = (EditText)findViewById(R.id.signature);
        signatureUse.setChecked(identity.getSignatureUse());
        signatureUse.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
            public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                if (isChecked) {
                    signatureLayout.setVisibility(View.VISIBLE);
                    signatureView.setText(identity.getSignature());
                } else {
                    signatureLayout.setVisibility(View.GONE);
                }
            }
        });

        if (signatureUse.isChecked()) {
            signatureView.setText(identity.getSignature());
        } else {
            signatureLayout.setVisibility(View.GONE);
        }
    }

    private void saveIdentity() {

        identity.setDescription(descriptionView.getText().toString());
        identity.setEmail(emailView.getText().toString());
        //      identity.setAlwaysBcc(mAccountAlwaysBcc.getText().toString());
        identity.setName(nameView.getText().toString());
        identity.setSignatureUse(signatureUse.isChecked());
        identity.setSignature(signatureView.getText().toString());

        if (replyTo.getText().length() == 0) {
            identity.setReplyTo(null);
        } else {
            identity.setReplyTo(replyTo.getText().toString());
        }

        List<Identity> identities = account.getIdentities();
        if (identityIndex == -1) {
            identities.add(identity);
        } else {
            identities.remove(identityIndex);
            identities.add(identityIndex, identity);
        }

        account.save(Preferences.getPreferences(getApplication().getApplicationContext()));

        finish();
    }

    @Override
    public void onBackPressed() {
        saveIdentity();
        super.onBackPressed();
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putSerializable(EXTRA_IDENTITY, identity);
    }
}

File: app/ui/legacy/src/main/java/com/fsck/k9/activity/EditIdentity.kt
package com.fsck.k9.activity

import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.CheckBox
import android.widget.TextView
import androidx.core.view.isVisible
import com.fsck.k9.Account
import com.fsck.k9.Identity
import com.fsck.k9.Preferences
import com.fsck.k9.ui.R
import com.fsck.k9.ui.base.K9Activity

class EditIdentity : K9Activity() {
    private lateinit var account: Account
    private lateinit var identity: Identity

    private lateinit var description: TextView
    private lateinit var name: TextView
    private lateinit var email: TextView
    private lateinit var replyTo: TextView
    private lateinit var signatureUse: CheckBox
    private lateinit var signature: TextView
    private lateinit var signatureLayout: View

    private var identityIndex: Int = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setLayout(R.layout.edit_identity)
        setTitle(R.string.edit_identity_title)
        supportActionBar!!.setDisplayHomeAsUpEnabled(true)

        identityIndex = intent.getIntExtra(EXTRA_IDENTITY_INDEX, -1)
        val accountUuid = intent.getStringExtra(EXTRA_ACCOUNT) ?: error("Missing account UUID")
        account = Preferences.getPreferences(this).getAccount(accountUuid) ?: error("Couldn't find account")

        identity = when {
            savedInstanceState != null -> savedInstanceState.getParcelable(EXTRA_IDENTITY) ?: error("Missing state")
            identityIndex != -1 -> intent.getParcelableExtra(EXTRA_IDENTITY) ?: error("Missing argument")
            else -> Identity()
        }

        description = findViewById(R.id.description)
        name = findViewById(R.id.name)
        email = findViewById(R.id.email)
        replyTo = findViewById(R.id.reply_to)
        signatureUse = findViewById(R.id.signature_use)
        signature = findViewById(R.id.signature)
        signatureLayout = findViewById(R.id.signature_layout)

        description.text = identity.description
        name.text = identity.name
        email.text = identity.email
        replyTo.text = identity.replyTo

        signatureUse.isChecked = identity.signatureUse
        signatureUse.setOnCheckedChangeListener { _, isChecked ->
            if (isChecked) {
                signatureLayout.isVisible = true
                signature.text = identity.signature
            } else {
                signatureLayout.isVisible = false
            }
        }

        if (signatureUse.isChecked) {
            signature.text = identity.signature
        } else {
            signatureLayout.isVisible = false
        }
    }

    private fun saveIdentity() {
        identity = identity.copy(
            description = description.text.toString(),
            email = email.text.toString(),
            name = name.text.toString(),
            signatureUse = signatureUse.isChecked,
            signature = signature.text.toString(),
            replyTo = if (replyTo.text.isNotEmpty()) replyTo.text.toString() else null
        )

        val identities = account.identities
        if (identityIndex == -1) {
            identities.add(identity)
        } else {
            identities.removeAt(identityIndex)
            identities.add(identityIndex, identity)
        }

        Preferences.getPreferences(applicationContext).saveAccount(account)

        finish()
    }

    public override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putParcelable(EXTRA_IDENTITY, identity)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.edit_identity_menu, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home) {
            finish()
            return true
        } else if (item.itemId == R.id.edit_identity_save) {
            saveIdentity()
            finish()
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    companion object {
        const val EXTRA_IDENTITY = "com.fsck.k9.EditIdentity_identity"
        const val EXTRA_IDENTITY_INDEX = "com.fsck.k9.EditIdentity_identity_index"
        const val EXTRA_ACCOUNT = "com.fsck.k9.EditIdentity_account"
    }
}
